src/id128/id128.c

```diff
@@ -55,7 +64,24 @@ static int verb_invocation_id(int argc, char **argv, void *userdata) {
                 return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
                                        "Verb \"invocation-id\" cannot be combined with --app-specific=.");
 
-        r = sd_id128_get_invocation(&id);
+        switch (arg_invocation_mode) {
+        case INVOCATION_AUTO:
+                r = sd_id128_get_invocation(&id);
+                break;
+
+        case INVOCATION_SYSTEM:
+                r = get_invocation_from_keyring(&id);
+                if (r == 0) /* get_invocation_from_keyring() returns 0 when keyring does not contain invocation ID. */
+                        r = -ENXIO;
+                break;
+
+        case INVOCATION_USER:
+                r = get_invocation_from_environment(&id);
+                break;
```

Comment:
Your patch looks like a reasonable change to current code, but I'm not sure if the *current* code is correct. b3415f5daef49642be3d5f417b8880c078420ff7 described the three ways that the invocation id is passed. Later d3070fbdf6077d7da9dbafa198fff8dea712d2ff added `/run/systemd/unit/` in addition to xattrs, for use by journald. So I'd  expect that we'd still rely on the environment variable in containers. But this doesn't seem to work, because we unconditionally error out without ever getting to `secure_getenv()`.

As root in a container, without your patch:
```console
$ strace systemd-id128 invocation-id
...
request_key("user", "invocation_id", NULL, 0) = -1 EPERM (Operation not permitted)
writev(2, [{iov_base="Failed to get invocation-ID: Ope"..., iov_len=52}, {iov_base="\n", iov_len=1}], 2Failed to get invocation-ID: Operation not permitted
) = 53
...
exit_group(1)
$ echo $INVOCATION_ID
b6f814efd4154598a17da7ec626b0035
```

Maybe @poettering can confirm what the expected behaviour is in this case: my expectation is that `sd_id128_get_invocation` would return stuff from the environment variable.